wit-manifest.json

```diff
@@ -1,17 +1,7 @@
 [
     {
-        "commit": "471dfa250b87a16a461f583e2b7eed69293cef02",
+        "commit": "97d5f359cfff3f4c35d55ad5cb1f362b6826f980",
         "name": "soc-testsocket-sifive",
         "source": "git@github.com:sifive/soc-testsocket-sifive.git"
-    },
-    {
-        "commit": "84d8bb615feb2975d6c4f543726563ca71c4538c",
-        "name": "api-generator-sifive",
-        "source": "git@github.com:sifive/api-generator-sifive.git"
-    },
-    {
-        "commit": "1c38068e5338dce8594f3aef0466c6333f46a24d",
-        "name": "api-scala-sifive",
-        "source": "git@github.com:sifive/api-scala-sifive.git"
```

Comment:
I'm interested in your philosophy of handling dependencies (in general, not just Wit).

My take is that any package (A) that directly makes use of something defined in some other package should directly depend on that package (B), even if it (A) depends on some other package (C) which also depends on it (B). The reason for that is that the fact that B depends on C is _not part of the API of B_, and so it should be acceptable for B to remove that dependency some time in the future without considering that to be a backwards incompatible change.

By that logic, because block-pio-sifive directly makes use of things in api-generator-sifive and in api-scala-sifive ((Wake rules, Python scripts run manually outside of Wake), then block-pio-sifive _should_ directly depend on them.

On the flip side, if we do in fact consider soc-testsocket-sifive's dependencies on the other two packages to be part of soc-testsocket-sifive's API, then I think this is fine, but then we should consider it to be a backwards incompatible change if we remove either of them.

Reply:
You know, that is a good point. I've always thought it was weird for Scala projects that depend on FIRRTL also depending on the fact that FIRRTL depends on 3rd party libraries like json4s.

It's completely inconsistent, but I haven't generally thought of Wit dependencies in the same way, but I think you're right. In general if you use an API you should directly depend on the source of that API.

Reply:
I just thought of one example of a case where the transitive dependencies _are_ part of the API of the direct dependencies, and it's [duh](https://github.com/sifive/duh) itself, which is really a metapackage of other packages. For `duh`, I think it would be fine to only depend on `duh` despite using the other `duh-*` packages that `duh` wraps around, since the main purpose of the `duh` package is to provide a set of versions of `duh-*` packages which are known to be compatible with each other.

Note that this would not apply to the other, Node.js arbitrary packages that `duh` relies on, such as its command line argument parsing library.

Reply:
@richardxia @albertchen-sifive @jackkoenig 

My philosophy is purely pragmatic. The fact that a casual observer (e.g. me) has no way of telling whether any given wit dependency exists in a manifest because of a true, direct, source code API version dependency, or whether it is e.g. a old bug-fix override of some child's transitive dependency means that we have quite a rat's nest of a dependency graph where parent modules are for whatever reason often pointing to grandchild dependency hashes that are months older than their childrens' currently pointed-to hashes. Personally I find this extremely confusing, especially since everything in manifests is represented as hashes even in the cases where versioned releases exist.

Coming at it from the other side, there is (IMO) no future in which `rocket-chip` removes its dependency on `api-generator-sifive` or `api-chisel3-sifive`, and I guess I would argue by analogy that the same applies to `soc-testsocket-sifive` and thus to this package too. There is simply, practically, absolutely no way "for [rocket-chip] to remove [the chisel] dependency some time in the future without considering that to be a backwards incompatible change." While I might be wrong about the directness of some dependencies in this repo specifically  w.r.t. to wake rules, I think in general the pattern I am describing is common for the Scala code and the upgrade pain/cognitive burden of the the more tightly coupled dependency graph is currently significant.

All that said, you guys are more responsible for this domain than I am at the moment, so I happy to concede to your philosophic consensus and will update this PR accordingly.

Reply:
I also see it as "if I'm putting the dependency explicitly in the top level repo, it means I don't want the one that I get automatically, for some reason".

Reply:
**Note: I am no longer a stakeholder of this repository, so take everything that I am saying with a grain of salt.**

> The fact that a casual observer (e.g. me) has no way of telling whether any given wit dependency exists in a manifest because of a true, direct, source code API version dependency, or whether it is e.g. a old bug-fix override of some child's transitive dependency means that we have quite a rat's nest of a dependency graph where parent modules are for whatever reason often pointing to grandchild dependency hashes that are months older than their childrens' currently pointed-to hashes. Personally I find this extremely confusing, especially since everything in manifests is represented as hashes even in the cases where versioned releases exist.

I completely agree with this, which is why I have opted for documenting the latter case with the comments that describe in detail when a transitive dependency is overridden. I can't find a good example of this in any of our public repositories, but I can point you to an internal repository in a private message.

That said, I'm not sure I understand how removing true dependencies from the wit manifest solves this problem, since you will still have the issue of having overridden transitive dependencies being present in wit-manifest.json. Is your philosophy that this is futile, so we might as well minimize the number of things written down in the wit-manifest.json file?

That said, I don't think I have an understanding of what your approach is, since I don't feel that I would be able to emulate what you would do in any given situation. Could you describe what your approach is independently of mine? I don't want to argue in favor of my approach without fully understanding yours, since we may actually have similar beliefs.

> Coming at it from the other side, there is (IMO) no future in which rocket-chip removes its dependency on api-generator-sifive

Hmm, I've had ideas in my head of splitting up api-generator-sifive, since it's a bit too much of an arbitrary grab bag of code, and sometimes I want to bump one piece of it without bumping the whole thing (e.g. I want the latest software driver flow fixes but not an update to rocket-chip). In any case, it's not a huge deal to have to deal with this when and if it actually happens.

> There is simply, practically, absolutely no way "for [rocket-chip] to remove [the chisel] dependency some time in the future without considering that to be a backwards incompatible change." 

Please keep in mind that api-chisel3-sifive is not the same thing as chisel3, so even if rocket-chip probably wouldn't ever get rid of its chisel3 dependency, I would find it much more realistic that rocket-chip would possibly get rid of its api-chisel3-sifive dependency, given that 1) it's been in there for less than a year and 2) it's not even in the git submodule list, since it's only used in wit/wake-based flows.

> we have quite a rat's nest of a dependency graph where parent modules are for whatever reason often pointing to grandchild dependency hashes that are months older than their childrens' currently pointed-to hashes

I don't actually view this as a problem per se. Let's ignore the problem of hashes being difficult for humans to understand, which I completely agree with, and for the sake of this argument, let's assume that we were using real version numbers (e.g. 1.3.11). Given a situation where A depends on B and C, and B depends on C: I don't think it's a bug for a A to depend on C v1.0.0 while B depends on C v1.2.0. Recall that wit dependencies describe "minimum" dependencies, so A depending on C v1.0.0 really means A depends on C >= v1.0.0. This makes it easier to remove A's dependency on B in the future.

---

Anyway, just to reiterate, I don't think I have much stake in the game with block-pio-sifive, since it's not a dependency of anything, as it is meant to be an example and not a library, so any decisions made about how to specify its dependencies are less of a concern. I _will_, however, probably feel more strongly about how dependencies are specified in key "library" repositories such as api-generator-sifive and rocket-chip.

Reply:
>  I can't find a good example of this in any of our public repositories, but I can point you to an internal repository in a private message.

I did not encounter any comments in the manifests involved in this bumping exercise, but perhaps that policy would be a sufficient compromise.

>  Could you describe what your approach is independently of mine?

If a dependency [rocket-chip] comprises mostly/entirely code written in one framework [chisel], then a direct dependency on the grandchild framework adds no value over the direct dependency on the user code written that framework. So I guess, yes, I am willing to go with "futile" in that case. But I'll admit claims of this sort are subject to my knowledge of the content of source codes of various downstream things which is not omniscient nor necessarily easily replicatable.

>  I don't think it's a bug for a A to depend on C v1.0.0 while B depends on C v1.2.0. 

I understand this is not a bug, but it is particularly confusing in the current situation (hashes, no comments).

Reply:
> I also see it as "if I'm putting the dependency explicitly in the top level repo, it means I don't want the one that I get automatically, for some reason".

@mwachs5 I think the problem is that since there is ultimately never a "top-level" repo that remains such for long, any such explicit top-level overrides will end up sucked deeper and deeper into the dependency chain with no explanation as to their presence, absent a comment.

Reply:
@hcook i think i am in violent agreement with what you are saying

Reply:
Maybe my final policy question is: is the goal to intentionally leave the pinned version at the lowest known version which is API-compatible? Like, in `soc-testsocket-sifive` right now, I am bumping `rocket-chip` explicitly to get the new chisel version. But (afaik) the new chisel was not required directly in `soc-testsocket-sifive` nor did it break any API compatibility there; so should I explicitly leave it at a lower version there despite knowing I am bumping `rocket-chip`'s transitive dependency past it?

Reply:
> Maybe my final policy question is: is the goal to intentionally leave the pinned version at the lowest known version which is API-compatible?

If it isn't too much of a hassle, then I think yes. For a pragmatic reason, at the moment there is no way using Wit to pin to an _older_ version of a transitive dependency. We actually ran into a problem several months ago where we accidentally introduced a bug in the latest api-scala-sifive, and had "opportunistically" bumped the version of api-scala-sifive used by some repo, even though we didn't actually need any of the new features from the latest version. Because of this, suddenly all of our top-level projects were "infected" by being pointed at the latest version of api-scala-sifive, and the only way to fix it was to explicitly introduce a revert commit in api-scala-sifive in master while we separately spent time to actually fix the bug.

Reply:
Hm, that would make sense to me if thats the idea behind the dependencies -- actually saying the dependency vs "thing I am guaranteeing to provide" (which is how the rocket-chip/firrtl relationship is sort of now)

Reply:
Sorry, I'm not trying to be obstinate, but I am having a really hard time understanding how not relying on the transitive dependency produces beneficial behavior. Just sticking to my specific example, `soc-testsocket-sifive` now points at a version of `rocket-chip` that requires `chisel3/v3.3`. `soc-testsocket-sifive` does use chisel APIs itself, but `v3.3` was backwards compatible enough that none of them broke. But what good does leaving `soc-testsocket-sifive`'s dep pointing at `v3.old` do? The transitive dependency is still real, if you manually check out `v3.old`, the rocket-chip in your workspace will not build. So if there is a bug in `chisel/v3.3` that rocket-chip has pulled in, how does knowing you could go back to `chisel/v3.old` help you?

Reply:
I'm not saying that you can manually check out `chisel/v3.old`, but if you removed the dependency on rocket-chip altogether (and maybe depending on some other Chisel-implemented core), then you still have specific requirements on the version of Chisel that soc-testsocket-sifive depends on.

Let me specifically call out another aspect of this that is problematic: Ignoring the Attachable interface, when using block-pio-sifive in a real SoC, block-pio-sifive _does not_ actually need the TestSocket. The TestSocket is really just a vehicle for testing a block on its own, but when integrating a block into a larger project, it shouldn't actually require that the TestSocket is API-compatible. This has _actually_ caused a problem for us in the past, where an IP block was onboarded with the TestSocket, and the block itself didn't depend on a newer rocket-chip, but when integrated into a larger SoC, the newer version of rocket-chip (and all of rocket-chip's dependencies) were pulled into that larger SoC, which caused a bunch of unnecessary compilation errors, since the TestSocket did not actually need to be part of the compilation of that SoC.

I agree that the transitive dependency is real for any given commit, but what if the direct dependency changes in the future?

Just keep in mind that I am talking about dependencies in general, and don't mean to specifically talk about chisel3, since I think chisel3 is a bit of a special case.

Reply:
Just going to jot this down for future reference. @terpstra and I had a chat here, and he made the point that if you have package B depend on package A and if package B exposes a public API that makes use of package A's APIs/data types, then using package B fundamentally relies on package A, and therefore you should be able to assume that package B will pull in package A.

Concretely: because rocket-chip.git depends on chisel3.git and rocket-chip.git exposes public interfaces involving chisel3 types, chisel3's types are part of rocket-chip's public API.

I buy this argument, and I think this provides a more consistent philosophy that captures a bit of both of the two opposing views above.